src/bin/ion/commands/beta/generate/model.rs

```diff
@@ -0,0 +1,268 @@
+    // Represents a top level scalar type definition which also has a name attached to it.
+    WrappedScalar(WrappedScalar),
+    // Represents a nested/inline scalar value (e.g. a string or integer or user defined type)
+    Scalar(Scalar),
+    // A series of zero or more values whose type is described by the nested `element_type`
+    Sequence(Sequence),
+    // A collection of field name/value pairs (e.g. a map)
+    Structure(Structure),
+}
+
+impl AbstractDataType {
+    #![allow(dead_code)]
+    pub fn doc_comment(&self) -> Option<String> {
+        match self {
+            AbstractDataType::WrappedScalar(WrappedScalar { doc_comment, .. }) => {
+                doc_comment.to_owned()
+            }
+            AbstractDataType::Scalar(Scalar { doc_comment, .. }) => doc_comment.to_owned(),
+            AbstractDataType::Sequence(Sequence { doc_comment, .. }) => {
+                Some(doc_comment.to_string())
+            }
+            AbstractDataType::Structure(Structure { doc_comment, .. }) => {
+                Some(doc_comment.to_string())
+            }
+        }
+    }
+
+    pub fn fully_qualified_name(&self) -> FullyQualifiedTypeName {
+        match self {
+            AbstractDataType::WrappedScalar(w) => w.fully_qualified_type_name().to_owned(),
+            AbstractDataType::Scalar(s) => s.name.to_owned(),
+            AbstractDataType::Sequence(seq) => seq.name.to_owned(),
+            AbstractDataType::Structure(structure) => structure.name.to_owned(),
+        }
+    }
+}
+
+/// Represents a nested/inline scalar type (e.g. a string or integer or user defined type)
+#[allow(dead_code)]
+#[derive(Debug, Clone, PartialEq, Serialize)]
+pub struct Scalar {
+    // Represents the fully qualified name for this data model
+    // For any nested/inline scalar type this would be the name of that type.
+    // e.g.
+    // ```
+    // type::{
+    //    name: foo,
+    //    type: list,
+    //    element: string // this is a nested scalar type
+    // }
+    // ```
+    // Corresponding `FullyQualifiedName` would be `vec!["String"]` and `scalar_type` would be `None`.
+    name: FullyQualifiedTypeName,
+    // Represents doc comment for the generated code
+    // If this is a top level scalar type definition then this will have `Some(doc_comment)`,
+    // Otherwise this is `None`.
+    doc_comment: Option<String>,
+    // Represents the source ISL type which can be used to get other constraints useful for this type.
+    // For example, getting the length of this sequence from `container_length` constraint or getting a `regex` value for string type.
+    // This will also be useful for `text` type to verify if this is a `string` or `symbol`.
+    // TODO: `IslType` does not implement `Serialize`, define a custom implementation or define methods on this field that returns values which could be serialized.
+    #[serde(skip_serializing)]
+    source: IslType,
+}
+
+/// Represents a top level scalar type definition
```

Comment:
I believe the idea here is actually that it represents a scalar type that is supposed to be nominally distinct from its "base" or Ion type. I don't think that there is necessarily a requirement that it be a top-level definition. If we eventually introduce e.g. `$codegen_name`, then any scalar type could have a name.

I think we should try to avoid having anything prescriptive about the ISL interpretation in this data model.